### ISSUE TITLE
Certain pipelines are double-caching

### ISSUE THREAD

**Issue by etrain** [amplab/keystone#258]
It appears that certain pipelines are caching intermediate state multiple times in separate storage areas, leading to more resource consumption than should be required. 

@shivaram noticed this in #247 but we've since observed it in a few places.

**Comment by tomerk** [comment]
Okay so it's not quite double-caching, it's a combination of two situations:

1. If the pipeline isn't fully constructed before training/applying to test data work may be repeated because the optimizer can't analyze the whole pipeline at once. It may be worth evaluating how intuitive the system is about encouraging the whole pipeline to be constructed before applying it to data.

2. If the pipeline is applied to training data, it is treated as test data and doesn't benefit from having already been processed/cached for training, leading work to be repeated. It may be worth adding some extra logic into KeystoneML to detect this and take advantage of any previously done work. This may have some unusual interactions w/ auto-caching that we'd need to look into.

**Comment by etrain** [comment]
Ok - can you provide a concrete example (e.g. code snippet) of what would cause a user to run into this scenario and an alternative way of writing it that will fix situation 1?

In terms of "pipeline applied to training data" case, we could either write code to handle this common case or we could think about how pipeline applications are represented to the system by the user. It seems like basically what's happening is the application is optimized in a narrower scope than "the whole application" - it is optimized whenever a `.apply` is called on a pipeline, rather than taking into account *all* times `apply` will be called on the pipeline. I'm not sure what the best way to support this notion is given our current programming model, but I think that's the jist of it. Does that make sense?

**Comment by tomerk** [comment]
Situation 1:

If a user does:
```scala
val featurizer = ...
val featurizedData = featurizer(trainData)
val model = estimator.fit(featurizedData, trainLabels)
val pipeline = featurizer andThen model
val predictions = pipeline(testData)
```

It should be replaced with:
```
val featurizer = ...
val pipeline = featurizer andThen (estimator, trainData, trainLabels)
val predictions = pipeline(testData)
```

As for the second, yeah that kinda makes sense, we'll need to revisit this question when I'm back.

**Comment by etrain** [comment]
Alright - thanks for the clarification. I guess this is what we created the `(estimator, data)` API for in the first place. I just removed the milestone from this issue since it doesn't look like a bug. But, let's leave the issue open so that we remember to brainstorm ways for capturing this behavior.

**Comment by shivaram** [comment]
I think combined with the fact that we know what data is materialized (e.g. if somebody uses the `Cacher` or does auto-caching) we should be able support a lot of incremental patterns.

**Comment by etrain** [comment]
Resolved by #277
